src/gwt/src/org/rstudio/studio/client/workbench/addins/AddinsCommandManager.java

```diff
@@ -104,26 +111,46 @@ public void execute(EditorKeyBindings bindings)
    private void registerBindings(final EditorKeyBindings bindings,
                                  final CommandWithArg<EditorKeyBindings> afterLoad)
    {
-      List<Pair<List<KeySequence>, CommandBinding>> commands =
-            new ArrayList<Pair<List<KeySequence>, CommandBinding>>();
-      
-      for (String id : bindings.iterableKeys())
-      {
-         List<KeySequence> keyList = bindings.get(id).getKeyBindings();
-         CommandBinding binding = new AddinCommandBinding(id);
-         commands.add(new Pair<List<KeySequence>, CommandBinding>(keyList, binding));
-      }
-      
-      KeyMap map = ShortcutManager.INSTANCE.getKeyMap(KeyMapType.ADDIN);
-      for (int i = 0; i < commands.size(); i++)
+      server_.getRAddins(false, new ServerRequestCallback<RAddins>()
```

Comment:
I would change this to be entirely event based (you populate the keymap only in response to an event issued from the server). Here is where you would emit the event from: https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/SessionRAddins.cpp#L255

This will cover re-indexing cases automatically and benefit from the file-based representation of the registry that we cache across sessions.